### PR TITLE
Refactor ReaDDy processing and all visualization

### LINKED ISSUE #21: Simultaneous visualization of sub-sampled cytosim and readdy simulations
### Feature Description
We want to simultaneously visualize sub-sampled readdy and cytosim simulations using appropriate colorization and metric plots

### Use Case
This will allow us to visually compare simulations across simulators and repeats.

### Solution
1. Each simularium trajectory should correspond to all repeats of cytosim and readdy for a particular velocity
2. This can be achieved by using `AgentData` append method in `simulariumio`. See `notebooks/create_simularium_outputs.py` for previous efforts in this direction.

### LINKED ISSUE #25: Import actin coordinates from tomography data and visualize and calculate metrics

### PR DESCRIPTION
**Pull request by blairlyons**
Time Estimate or Size
=======
x-large (sorry!)

Problem
=======
Visualization and post processing scripts were inefficiently organized and out of date. 

Solution
========
I refactored code and tested all of the notebooks in simulation.readdy and visualization modules.

## Type of change
Please delete options that are not relevant.

* Breaking change (fix or feature that would cause existing functionality to not work as expected)
* This change requires a documentation update

Keyfiles:
-----------------------
Main changes are in simulation.readdy and visualization modules, but I had to make some changes to analysis and other places to support these.

### CODE REVIEW

File: subcell_pipeline/analysis/compression_metrics/README.md [hunk 1]
```diff
@@ -2,6 +2,6 @@
 
 ## Metrics for comparing traces of compressed fibers
 
-Analysis combines compression simulations from Cytosim and Readdy and calculates various metrics to compare the compressed fibers.
+Analysis combines compression simulations from Cytosim and Readdy and calculates various compression metrics metrics to compare fibers.
```
Comment (jessicasyu) [inline subcell_pipeline/analysis/compression_metrics/README.md]:
```suggestion
Analysis combines compression simulations from Cytosim and Readdy and calculates various compression metrics to compare fibers.
```

File: subcell_pipeline/analysis/dimensionality_reduction/fiber_data.py [hunk 2]
```diff
@@ -104,9 +104,9 @@ def align_fibers(data: pd.DataFrame) -> None:
     data["zpos"] = all_aligned_fibers[:, 2]
 
 
-def align_fiber(coords: np.ndarray) -> np.ndarray:
+def align_fiber(coords: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
     """
-    Align an array of x, y, z coordinates along the positive y axis.
+    Align an array of x, y, z coordinates along the positive x axis.
```
Comment (mogres) [inline subcell_pipeline/analysis/dimensionality_reduction/fiber_data.py]: I think this is supposed to be the y axis?

Comment (jessicasyu) [inline subcell_pipeline/analysis/dimensionality_reduction/fiber_data.py]: Positive y axis is correct here. The function uses the yz plane and rotates that "loop" to align onto the y axis

File: subcell_pipeline/analysis/tomography_data/_analyze_actin_cme_tomography_data.py [hunk 3]
```diff
@@ -26,13 +26,20 @@
...
+
 from subcell_pipeline.analysis.tomography_data.tomography_data import (
     get_branched_tomography_data,
     get_unbranched_tomography_data,
     plot_tomography_data_by_dataset,
     sample_tomography_data,
 )
 
+# pixels to um
+TOMOGRAPHY_SCALE_FACTOR: float = 0.0006
```
Comment (jessicasyu) [inline subcell_pipeline/analysis/tomography_data/_analyze_actin_cme_tomography_data.py]: I would leave this as a configuration parameter instead of a constant since it isn't imported anywhere else and depends on the data set.

File: subcell_pipeline/analysis/tomography_data/tomography_data.py [hunk 4]
```diff
@@ -1,11 +1,41 @@
...
 from io_collection.keys.check_key import check_key
 from io_collection.load.load_dataframe import load_dataframe
+from io_collection.save.save_buffer import save_buffer_to_s3
 from io_collection.save.save_dataframe import save_dataframe
+from PIL import Image
+
+TOMOGRAPHY_SAMPLE_COLUMNS: list[str] = ["xpos", "ypos", "zpos"]
+
+
+def test_consecutive_segment_angles(polymer_trace: np.ndarray) -> np.bool_:
```
Comment (jessicasyu) [inline subcell_pipeline/analysis/tomography_data/tomography_data.py]: Can we have the return be a regular `bool`? I think you can just have the return be `np.all(dot_products > 0).item()` and it should work.

File: subcell_pipeline/analysis/tomography_data/tomography_data.py [hunk 5]
```diff
@@ -1,11 +1,41 @@
...
 
-SAMPLE_COLUMNS = ["xpos", "ypos", "zpos"]
+    Parameters
+    ----------
+    polymer_trace
+        A 2D array where each row is a point in 3D space.
+
+    Returns
+    -------
+    bool
```
Comment (jessicasyu) [inline subcell_pipeline/analysis/tomography_data/tomography_data.py]: nit: replace return type with `:` so you don't need to keep two type hints in sync (the autodoc will grab the one in the method signature)

File: subcell_pipeline/analysis/tomography_data/tomography_data.py [hunk 6]
```diff
@@ -272,4 +329,6 @@ def plot_tomography_data_by_dataset(data: pd.DataFrame) -> None:
             ax[1].plot(fiber["xpos"], fiber["zpos"], marker="o", ms=1, lw=1)
             ax[2].plot(fiber["ypos"], fiber["zpos"], marker="o", ms=1, lw=1)
 
-        plt.show()
+    plt.savefig(local_save_path)
+    image: np.ndarray = imageio.imread(local_save_path)
+    save_image_to_s3(bucket, output_key, image)
```
Comment (jessicasyu) [inline subcell_pipeline/analysis/tomography_data/tomography_data.py]: You can use `io_collection.save.save_figure` here to save the `fig` to a local path or S3 bucket. I think you want to move the figure saving into the loop (and update the output key) so you save each individual figure, rather than just the final figure.

File: subcell_pipeline/simulation/readdy/loader.py [hunk 7]
```diff
@@ -1,83 +1,96 @@
...
 
 import numpy as np
 import readdy
+from io_collection.keys.check_key import check_key
+from io_collection.load.load_pickle import load_pickle
+from io_collection.save.save_pickle import save_pickle
 from tqdm import tqdm
 
-from .readdy_data import FrameData, ParticleData, TopologyData
+from .data_structures import FrameData, ParticleData, TopologyData
```
Comment (jessicasyu) [inline subcell_pipeline/simulation/readdy/loader.py]: nit: Absolute instead of relative import

File: subcell_pipeline/visualization/_visualize_tomography_data.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
+# %% [markdown]
+# # Visualize actin CME tomography data
+
+# %% [markdown]
+"""
+Notebook contains steps for visualizing segmented tomography data for actin
+fibers using [Simularium](https://simularium.allencell.org/).
+
+- [Define visualization settings](#define-visualization-settings)
+- [Visualize tomography data](#visualize-tomography)
```
Comment (jessicasyu) [inline subcell_pipeline/visualization/_visualize_tomography_data.py]:
```suggestion
- [Visualize tomography data](#visualize-tomography-data)
```

File: subcell_pipeline/visualization/_visualize_cytosim_trajectories.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+# %%
+# Name of the simulation series
+compression_series_name: str = "COMPRESSION_VELOCITY"
+
+# List of condition file keys for each velocity
+compression_condition_keys: list[str] = ["0047", "0150", "0470", "1500"]
+
+# %%
+visualize_individual_cytosim_trajectories(
```
Comment (jessicasyu) [inline subcell_pipeline/visualization/_visualize_cytosim_trajectories.py:87]: I was able to run this successfully (just for the first three seeds of the `0047` condition) but the linkers seem to be misaligned with the actual fiber.

**Review by jessicasyu — COMMENTED** [review]
Looks great! Mostly minor things. The only major issue is with the Cytosim simularium trajectories, where the linkers seem to be off.

There's also some missing docstrings, but we can add those in a separate PR so this one doesn't keep getting bigger ;)

File: subcell_pipeline/simulation/readdy/_process_readdy_compression_simulations.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+# Number of timepoints
+n_timepoints = 200
+
+# Number of monomer points per fiber
+n_monomer_points = 200
+
+# Total number of steps for each condition
+total_steps: dict[str, int] = {
+    "0047": int(3.2e8),
```
Comment (mogres) [inline subcell_pipeline/simulation/readdy/_process_readdy_compression_simulations.py]: should these be calculated based on the velocity and the compression distance for consistency instead of being hard coded?

File: subcell_pipeline/simulation/readdy/loader.py [hunk 11]
```diff
@@ -1,83 +1,96 @@
...
+
     def __init__(
         self,
         h5_file_path: str,
         min_time_ix: int = 0,
         max_time_ix: int = -1,
         time_inc: int = 1,
         timestep: float = 100.0,
-        save_pickle_file: bool = False,
+        pickle_location: Optional[str] = None,
```
Comment (mogres) [inline subcell_pipeline/simulation/readdy/loader.py:62]: Does this function need documentation or the class attribute descriptions are sufficient?

File: pyproject.toml [hunk 12]
```diff
@@ -27,6 +27,7 @@ dependencies = [
     "setuptools>=70.0.0",
     "io-collection>=0.10.2",
     "python-dotenv>=1.0.1",
+    "imageio",
```
Comment (mogres) [inline pyproject.toml]: Would recommend adding packages using pdm by running `pdm add imageio`. This will also update the lockfile correctly for `pdm sync`

**Review by mogres — COMMENTED** [review]
Thanks for making the changes, the visualization updates are very helpful! The PR overall looks good to me. I was able to run the visualization notebooks and generate the simularium trajectories but ran into a couple of issues along the way.

1) In the individual readdy and cytosim visualization notebooks, I ran into an error due to the missing `h5py` package. After installing it using `pdm add h5py` the notebooks ran as expected.

2) I also ran into the issue Jess mentioned regarding the linkers in Cytosim trajectories. They seem to be offset from the main fiber.
image

File: subcell_pipeline/analysis/tomography_data/tomography_data.py (hunk 6 above)
Comment (blairlyons) [reply subcell_pipeline/analysis/tomography_data/tomography_data.py]: oh I tried to use save image but had to avoid it because I didn't want to depend on bioio for the OmeTiffWriter. It would help if this package was documented :)

Comment (jessicasyu) [reply subcell_pipeline/analysis/tomography_data/tomography_data.py]: Haha yes I've been slowly working on the docs :)

The default install of `io-collection` does not include the `bioio` install, so should be okay.

File: subcell_pipeline/simulation/readdy/loader.py (hunk 11 above)
Comment (blairlyons) [reply subcell_pipeline/simulation/readdy/loader.py:62]: since it's the constructor I usually just rely on the class documentation

File: subcell_pipeline/visualization/_visualize_cytosim_trajectories.py (hunk 9 above)
Comment (blairlyons) [reply subcell_pipeline/visualization/_visualize_cytosim_trajectories.py:87]: I forgot, I had to make a small change to simulariumio to fix this. It's in [PR](https://github.com/simularium/simulariumio/pull/188) now and will hopefully be released shortly

**Comment by blairlyons** [comment]
@jessicasyu @mogres I think I addressed everything, the Cytosim issue depends on [this PR](https://github.com/simularium/simulariumio/pull/188) getting merged. I tried fixing the mypy error in the analysis module without success, would be great if someone knows how to fix that.

File: subcell_pipeline/analysis/tomography_data/tomography_data.py [hunk 13]
```diff
@@ -329,6 +313,4 @@ def plot_tomography_data_by_dataset(
             ax[1].plot(fiber["xpos"], fiber["zpos"], marker="o", ms=1, lw=1)
             ax[2].plot(fiber["ypos"], fiber["zpos"], marker="o", ms=1, lw=1)
 
-    plt.savefig(local_save_path)
-    image: np.ndarray = imageio.imread(local_save_path)
-    save_image_to_s3(bucket, output_key, image)
+    save_figure(bucket, output_key, figure)
```
Comment (jessicasyu) [inline subcell_pipeline/analysis/tomography_data/tomography_data.py]: I think you want to save separate figures for each dataset?

File: .github/workflows/build.yml [hunk 14]
```diff
@@ -53,12 +53,6 @@ jobs:
         conda env update --file environment.yml --prune
         pdm sync
 
-    - name: Test with pytest
-      shell: bash -l {0}
-      run: |
-        pdm run pytest --cov --cov-report html
-        rm htmlcov/.gitignore
-
```
Comment (jessicasyu) [inline .github/workflows/build.yml:61]: I would keep the build workflow as is (it'll break the coverage badge otherwise). It just won't find any tests to run, which I think is fine.

File: .pdm-python [hunk 15]
```diff
@@ -0,0 +1 @@
+/Users/blairl/Documents/Dev/subcell-pipeline/.venv/bin/python
```
Comment (jessicasyu) [inline .pdm-python]: We shouldn't commit this file -- probably add to the `.gitignore`?

File: subcell_pipeline/simulation/readdy/parser.py [hunk 16]
```diff
@@ -205,14 +209,23 @@ def parse_readdy_simulation_single_fiber_trajectory(
     return dataframe
 
 
+def round_2_sig_figs(x: float) -> int:
```
Comment (jessicasyu) [inline subcell_pipeline/simulation/readdy/parser.py]: Needs docstrings (or make private with `_` prefix and they won't show up in the docs)

**Review by jessicasyu — APPROVED** [review]
Looks great! Just a couple more things but good to merge after that. I can take a final pass through everything once we get the PRs all merged.

**Comment by jessicasyu** [comment]
@blairlyons @mogres I believe that last mypy issues is fixed in #61

**Review by mogres — APPROVED** [review]

File: .github/workflows/build.yml (hunk 14 above)
Comment (blairlyons) [reply .github/workflows/build.yml:61]: it breaks the build otherwise. We can add one test that passes, or maybe the coverage badge should be removed since we know it's zero, unfortunately

File: .pdm-python (hunk 15 above)
Comment (blairlyons) [reply .pdm-python]: whoops

File: .github/workflows/build.yml (hunk 14 above)
Comment (jessicasyu) [reply .github/workflows/build.yml:61]: I can refactor the existing test so it passes? It doesn't need to be in this PR

Comment (blairlyons) [reply .github/workflows/build.yml:61]: can I leave it like this for this PR and you can pull this back in with the fixed test?
